### ISSUE TITLE
Wrappers for subroutines with assumed-shape arrays to improve interoperability with other languages?

### ISSUE THREAD

**Issue by jishnub** [SHTOOLS/SHTOOLS#177]
It is difficult to call Fortran subroutines with assumed-shaped arrays. I wonder if some wrappers can be added that explicitly specify the dimensions in order to make it easier to call shtools from other languages, eg Julia?

**Comment by MarkWieczorek** [comment]
Could you provide an example of how this would work? I have never tried to use these routines with anything besides python.

**Comment by jishnub** [comment]
Here is an example of a wrapper that I wrote
```
module shtools_wrapper

use, intrinsic :: iso_c_binding
use SHTOOLS
implicit none

contains 

subroutine wigner3j_wrapper(w3j,len,j2, j3, m1, m2, m3, exitstatus) bind(C, name="wigner3j_wrapper")
    integer(kind=C_INT32_T), intent(in)  :: len,j2,j3,m1,m2,m3
    real(kind=C_DOUBLE), intent(out) :: w3j(len)
    integer(kind=C_INT32_T) :: jmin,jmax
    integer(kind=C_INT32_T), intent(out), optional :: exitstatus

    call Wigner3j(w3j, jmin, jmax, j2, j3, m1, m2, m3, exitstatus=exitstatus)
end subroutine wigner3j_wrapper

end module shtools_wrapper
```

This can then be compiled to a shared library and called from Julia directly as 
```
function wig3j(j2,j3,m2,m3)
	
	m2,m3 = Int32(m2),Int32(m3)
	m1 = Int32(m2 + m3)

	j2,j3 = Int32(j2),Int32(j3)
	len = Int32(j2+j3+1)

	exitstatus = zero(Int32)

	w3j = zeros(Float64,len)

	ccall((:wigner3j_wrapper,"./shtools_wrapper.so"),Cvoid,
		(Ref{Float64}, 	#w3j
			Ref{Int32},	#len
			# Ref{Int32},	#jmin
			# Ref{Int32},	#jmax
			Ref{Int32},	#j2
			Ref{Int32},	#j3
			Ref{Int32},	#m1
			Ref{Int32},	#m2
			Ref{Int32},	#m3
			Ref{Int32}),#exitstatus
		w3j,len, j2, j3, m1, m2,m3, exitstatus)
	return w3j
end
```

The main difference is that I have specified the length of w3j explicitly and used C bindings. This ensures that the parameters are explicitly used as specified, otherwise the gfortran compiler seems to introduce hidden parameters corresponding to the size of w3j and a call to the subroutine fails.

**Comment by MarkWieczorek** [comment]
I can see why this would be useful, but I am not going to have the time to work on this for some time... If someone wants to develop this, that would be great :)  Perhaps the first step would be to put in a PR that sets up everything to work with a single routine.

**Comment by rainbowsend** [comment]
@jishnub can you use the  [cWrapper](https://github.com/rainbowsend/SHTOOLS/blob/fa7f0811f81c46b69abc7aa9c38ad2c8a46472d8/src/cWrapper.f95) to call shtools functions from julia?

The wrapper includes all functions that are in SHTOOLS.f95. For example, the function you have picked above:

```Fortran
    subroutine cWigner3j(w3j,jmin,jmax,j2,j3,m1,m2,m3,exitstatus)  bind(c, name="Wigner3j")
        use, intrinsic :: iso_c_binding
        use shtools, only: Wigner3j
        implicit none
        integer(kind=c_int), value,intent(in) :: j3
        integer(kind=c_int), value,intent(in) :: j2
        integer(kind=c_int), value,intent(in) :: m1
        integer(kind=c_int), value,intent(in) :: m2
        integer(kind=c_int), value,intent(in) :: m3
        integer(kind=c_int), intent(out) :: jmin
        integer(kind=c_int), intent(out) :: jmax
        real(kind=c_double), dimension(j2+j3+1),intent(out) :: w3j
        integer(kind=c_int), optional,intent(out) :: exitstatus
        call Wigner3j(w3j,jmin,jmax,j2,j3,m1,m2,m3,exitstatus=exitstatus)
    end subroutine cWigner3j
```

**Comment by jishnub** [comment]
This looks great, exactly what I wanted. Will try it out

**Comment by MarkWieczorek** [comment]
This was just included in the 4.7.1 release, and a working example can be found in `examples/cpp`.

I am going to close this issue, but please feel free to re-open it if you have any problems. The C binded functions are still a work in progress, and we understand that there is no documentation at this point. This will hopeful be resolved for the next minor release (4.8).
